app/gui2/src/util/ast/aliasAnalysis.ts

```diff
@@ -0,0 +1,276 @@
+
+  public readonly aliases: ObjectKeyedMap<ContentRange, ObjectKeyedSet<ContentRange>> =
+    new ObjectKeyedMap()
+
+  /**
+   * @param code text representation of the code.
+   * @param ast AST representation of the code. If not provided, it will be parsed from the text.
+   */
+  constructor(
+    private readonly code: string,
+    ast?: Tree,
+  ) {
+    this.ast = ast ?? parseEnso(code)
+    this.rootScope = new Scope(parsedTreeOrTokenRange(this.ast))
+    this.scopes = new NonEmptyStack(this.rootScope)
+  }
+
+  /** Invoke the given function in a new temporary scope. */
+  withNewScopeOver(nodeOrRange: ContentRange | Tree | Token, f: () => undefined) {
+    const range = parsedTreeOrTokenRange(nodeOrRange)
+    const scope = new Scope(range, this.scopes.top)
+    this.scopes.withPushed(scope, f)
+  }
+
+  /** Execute the given function while the given context is temporarily pushed to the stack. */
+  withContext(context: Context, f: () => void) {
+    return this.contexts.withPushed(context, f)
+  }
+
+  /** Marks given token as a binding, i.e. a definition of a new variable. */
+  bind(token: Token) {
+    const scope = this.scopes.top
+    const identifier = readTokenSpan(token, this.code)
+    const range = parsedTreeOrTokenRange(token)
+    log(`Binding ${identifier}@[${range}]`)
+    scope.bindings.set(identifier, token)
+    assert(!this.aliases.has(range), `Token at ${range} is already bound.`)
+    this.aliases.set(range, new ObjectKeyedSet())
+  }
+
+  addConnection(source: Token, target: Token) {
+    const sourceRange = parsedTreeOrTokenRange(source)
+    const targetRange = parsedTreeOrTokenRange(target)
+    const targets = this.aliases.get(sourceRange)
+    if (targets != null) {
+      log(
+        `Usage of ${readTokenSpan(
+          source,
+          this.code,
+        )}@[${targetRange}] resolved to [${sourceRange}]`,
+      )
+      targets.add(targetRange)
+    } else {
+      console.warn(`No targets found for source range ${sourceRange}`)
+    }
+  }
+
+  /** Marks given token as a usage, i.e. a reference to an existing variable. */
+  use(token: Token) {
+    const identifier = readTokenSpan(token, this.code)
+    const range = parsedTreeOrTokenRange(token)
+
+    for (const scope of this.scopes.valuesFromTop()) {
+      const binding = scope.bindings.get(identifier)
+      if (binding != null) {
+        this.addConnection(binding, token)
+        return
+      }
+    }
+
+    log(`Usage of ${identifier}@[${range}] is unresolved.`)
+    this.unresolvedSymbols.add(range)
+  }
+
+  /** Method that processes a single AST node. */
+  public process() {
+    this.processNode(this.ast)
+  }
+
+  processToken(token?: Token): void {
+    if (token == null) {
+      return
+    }
+
+    const repr = readTokenSpan(token, this.code)
+
+    // We require first character to be lowercase, to avoid working with type names.
+    if (repr.length > 0 && repr.charAt(0) !== repr.charAt(0).toUpperCase()) {
```

Comment:
Is this "uppercase means a type" semantic actually part of the language? Also, we may as well collect used types as "things used within the scope". So maybe this `if` could simply be removed.

Reply:
> Is this "uppercase means a type" semantic actually part of the language?

It seems to be part of language, at least in https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/03.%20Code%20format%20and%20layout.md#type-identifiers
Anyway, we need something to tell apart pattern-matching variable against type and binding a variable. 

>  Also, we may as well collect used types as "things used within the scope". So maybe this `if` could simply be removed.

Could be done, but then more logic for handling type definitions would be needed. This should be tracked as a separate piece of work, likely along with some use-case.